src/zwarpedarray.jl

```diff
@@ -1,43 +1,79 @@
-mutable struct ZWarpedArray{TO,TI,N} <: CachedSeries2D{TO,TI,N}
-    parent::AbstractArray{TI,N}
+mutable struct ZWarpedArray{TO,TI,N,NC,AA<:AbstractArray} <: AbstractCachedArray{TO,TI,N,NC,AA}
+    parent::AA
     tfms::Vector{CoordinateTransformations.Transformation}
-    cached::Array{TO,2}
-    cache_idxs::Tuple
+    cached::Array{TO,NC}
+    current_I::Tuple
     correct_bias::Bool
     sqrt_tfm::Bool
 end
 
 cache(A::ZWarpedArray) = A.cached
-cache_idxs(A::ZWarpedArray) = A.cache_idxs
+current_I(A::ZWarpedArray) = A.current_I
+set_I!(A::ZWarpedArray, inds) = A.current_I = inds
 
-function ZWarpedArray(img::Array34{T}, tfms, out_type=Float64; correct_bias=true, sqrt_tfm=false) where {T}
+function ZWarpedArray(img::AbstractArray{T, N}, tfms, out_type=Float64, nd_cache::Int=2; correct_bias=true, sqrt_tfm=false) where {T, N}
+    if !(2 < ndims(img) < 5)
+        error("Only 3D and 4D images are supported")
+    end
     if size(img,3) !== length(tfms)
         error("Input image size in the Z-slice dimension (3) does not match the number of transforms provided")
     end
+    if nd_cache < 2 || nd_cache > 3
+        error("nd_cache must be 2 or 3")
+    end
     if sqrt_tfm && !correct_bias
         warn("Square root transformation will yield incorrect results if bias was not subtracted first")
     end
-    za = ZWarpedArray{out_type, T, ndims(img)}(img, tfms, zeros(out_type, size(img)[1:2]...), (ones(ndims(img)-2)...), correct_bias, sqrt_tfm)
-    update_cache!(za, (ones(Int, ndims(img)-2)...))
+    if correct_bias && !in(T, (Normed{UInt16, 16}, UInt16))
+        warn("Unable to guess the camera bias value for element type $T.  Defaulting to 100/(2^16)")
+    end
+    cache_rngs = axes(img)[nd_cache+1:ndims(img)]
+    ci = map(first, cache_rngs)
+    cached = similar(view(img, axes(img)[1:nd_cache]..., ci...), out_type)
+    za = ZWarpedArray{out_type, T, ndims(img), nd_cache, typeof(img)}(img, tfms, cached, (ci...,), correct_bias, sqrt_tfm)
+    update_cache!(za, (ci...,))
     return za
 end
 
-function update_cache!(A::ZWarpedArray{TO, TI, N}, inds::NTuple{N2, Int}) where {TO, TI, N, N2}
-    pslice = view(A.parent, :, :, inds...)
-    tfm = A.tfms[inds[1]] #get transform for this slice
+function _update_cache!(c::AbstractArray{T,2}, parent, tfm::CoordinateTransformations.Transformation, pp::Function) where {T}
+    c .= pp.(warp_and_resample(parent, tfm)) #TODO: do this in-place
+    return nothing
+end
+
+function _update_cache!(c::AbstractArray{T,3}, parent, tfms, pp::Function) where {T}
+    for (i,v) in enumerate(axes(c,3))
+        _update_cache!(view(c, :, :, v), view(parent, :, :, v), tfms[i], pp)
+    end
+    return nothing
+end
+
+function biasval(t::Type)
+    native = UInt16(100)
```

Comment:
Obviously it would be great to support multiple cameras, which may choose different values here, but this is fine for now.